SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditorPackageAssembly.cpp

```diff
@@ -0,0 +1,242 @@
+		}
+		if (SpatialGDKSettings->IsSimulatedPlayersEnabled())
+		{
+			Steps.Enqueue(EPackageAssemblyStep::BUILD_SIMULATED_PLAYERS);
+		}
+		Steps.Enqueue(EPackageAssemblyStep::UPLOAD_ASSEMBLY);
+
+		AsyncTask(ENamedThreads::GameThread, [this]()
+		{
+			this->ShowTaskStartedNotification(TEXT("Building Assembly"));
+			this->NextStep();
+		});
+	}
+}
+
+bool FSpatialGDKPackageAssembly::CanBuild() const
+{
+	return Steps.IsEmpty();
+}
+
+bool FSpatialGDKPackageAssembly::NextStep()
+{
+	bool HasMoreSteps = false;
+	EPackageAssemblyStep Target = EPackageAssemblyStep::NONE;
+	if (Steps.Dequeue(Target))
+	{
+		HasMoreSteps = true;
+		switch(Target)
+		{
+		case EPackageAssemblyStep::BUILD_SERVER:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FString::Printf(TEXT("%sServer"), FApp::GetProjectName()), Linux, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_CLIENT:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FApp::GetProjectName(), Win64, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::BUILD_SIMULATED_PLAYERS:
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->BuildAssembly(FString::Printf(TEXT("%sSimulatedPlayer"), FApp::GetProjectName()), Linux, AssemblyDetailsPtr->Configuration, TEXT(""));
+			});
+			break;
+		case EPackageAssemblyStep::UPLOAD_ASSEMBLY:
+		{
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				this->AssemblyDetailsPtr->Upload(*this);
+			});
+		}
+		break;
+		}
+	}
+	return HasMoreSteps;
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCompleted(int32 TaskResult)
+{
+	if (TaskResult == 0)
+	{
+		if (!NextStep())
+		{
+			AsyncTask(ENamedThreads::GameThread, [this]()
+			{
+				FString NotificationMessage = FString::Printf(TEXT("Assembly successfully uploaded to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+				this->ShowTaskEndedNotification(NotificationMessage, SNotificationItem::CS_Success);
+				OnSuccess.ExecuteIfBound();
+			});
+		}
+	}
+	else
+	{
+		FString NotificationMessage = FString::Printf(TEXT("Failed assembly upload to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+		AsyncTask(ENamedThreads::GameThread, [this]() { this->ShowTaskEndedNotification(TEXT("Assembly Failed"), SNotificationItem::CS_Fail); });
+		Steps.Empty();
+	}
+}
+
+void FSpatialGDKPackageAssembly::OnTaskOutput(FString Message)
+{
+	UE_LOG(LogSpatialGDKEditorPackageAssembly, Display, TEXT("%s"), *Message);
+}
+
+void FSpatialGDKPackageAssembly::OnTaskCanceled()
+{
+	Steps.Empty();
+	FString NotificationMessage = FString::Printf(TEXT("Cancelled assembly upload to project: %s"), *FSpatialGDKServicesModule::GetProjectName());
+	AsyncTask(ENamedThreads::GameThread, [this, NotificationMessage]()
+	{
+		this->ShowTaskEndedNotification(NotificationMessage, SNotificationItem::CS_Fail);
+		this->AssemblyDetailsPtr.Reset();
```

Comment:
Wonder if it might be neat to reset AssemblyDetailsPtr in OnTaskCompleted when we either receive a bad TaskResult or have finished all steps.